zerver/migrations/0377_message_edit_history_format.py

```diff
@@ -0,0 +1,165 @@
+import time
+from typing import List, Optional
+
+import orjson
+from django.db import migrations, transaction
+from django.db.backends.postgresql.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from django.db.models import Min, Model
+from typing_extensions import TypedDict
+
+BATCH_SIZE = 1000
+STREAM = 2
+
+# Legacy TypedDict has "prev_topic" field for any edit_history entries that
+# were saved to the database after the legacy "prev_subject" field stopped
+# being written to the database in the pre-migration commit.
+class LegacyEditHistoryEvent(TypedDict, total=False):
+    user_id: int
+    timestamp: int
+    prev_stream: int
+    prev_subject: str
+    prev_topic: str
+    prev_content: str
+    prev_rendered_content: Optional[str]
+    prev_rendered_content_version: Optional[int]
+
+
+class EditHistoryEvent(TypedDict, total=False):
+    user_id: int
+    timestamp: int
+    prev_stream: int
+    stream: int
+    prev_topic: str
+    topic: str
+    prev_content: str
+    prev_rendered_content: Optional[str]
+    prev_rendered_content_version: Optional[int]
+
+
+def backfill_message_edit_history_chunk(first_id: int, last_id: int, message_model: Model) -> None:
+    """
+    Migrate edit history events for the messages in the provided range to:
+    * Rename prev_subject => prev_topic.
+    * Provide topic and stream fields with the current values.
+    """
+    messages = (
+        message_model.objects.select_for_update()
+        .only(
+            "recipient",
+            "recipient__type",
+            "recipient__type_id",
+            "subject",
+            "edit_history",
+        )
+        .filter(edit_history__isnull=False, id__range=(first_id, last_id))
+    )
```

Comment:
I think it'd be most readable to put this inside `with transaction.atomic`; structurally, it's no different, because this query is only lazily evaluated below, but that's very much not obvious to a reader not familiar with Django's lazy evaluation model :).

Reply:
Yeah, that makes sense. My putting the query first was from using the `select_for_update` example from the Django docs as a model for my migration structure - https://docs.djangoproject.com/en/4.0/ref/models/querysets/#select-for-update.